src/http-client/fetch-client.ts

```diff
@@ -20,6 +21,10 @@ export class FetchClient implements HTTPClient {
       method,
       headers: { ...requestHeaders, "Content-Type": "application/json" },
       body: JSON.stringify(data),
+    }).catch((error) => {
+      throw new NetworkError("The network connection encountered a problem.", {
+        cause: error,
+      });
     });
```

Comment:
this is too generic.

There's no guarantee this a Network error. See [fetch docs](https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions) for other possiblities.

Let's delete this block and let the client class sort it out.

Reply:
Ah, makes sense. I had reviewed the fetch docs, but I think got hung up on this line at the top

> A [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch) promise only rejects when a network error is encountered (which is usually when there's a permissions issue or similar).

So once the request is made, it's definitely a network error, but while you are generating a request it can still throw other errors 👍🏻 

We can check for `error instance of DOMException` and `error.name === "NetworkError"`.

Note that `DOMException` only makes sense in the context of web APIs (e.g. fetch). That's probably more than the Client should have to worry about. Counter argument is that since developers can provide their own HTTPClient we would have better error handling if the developer fails to handles the errors themselves. Favoring better error handling in the user-provided http client, I'll push the error handling to the Client class.

Reply:
The distinction is between the Promise for `fetch()` rejecting and the Promise for `response.text()` rejecting. The former indicates one of several errors like bad headers, for example. The latter will always be a network error.

It is straightforward to make the distinction inside of the HTTPClient. both the bad-request errors and network errors are thrown as `TypeError` we cannot make the distinction outside of the HTTPClient.

I think we should update FetchClient to catch and re-throw a `NetworkError` on `response.text()` and catch and re-throw a `ProtocolError` on the `fetch()` call.

Reply:
bad headers do not fit with our definition of network error. we should huddle on this to get on the same page for terminology.

Reply:
discussed offline. moving finer distinctions to P2